scarlet/source.py

```diff
@@ -213,42 +253,198 @@ def __init__(
+            )
+            im = prox_monotonic(im, 0).reshape(im.shape)
+
+        # truncate morph at thresh * bg_rms
+        threshold = detect_std * thresh
+        morph, bbox = init.trim_morphology(center_index, im, bg_thresh=threshold)
+
+        # normalize to unity at peak pixel for the imposed normalization
+        if morph.sum() > 0:
+            morph /= morph.max()
+        else:
+            morph = CenterOnConstraint(tiny=1)(morph, 0)
+            msg = f"No flux in morphology model for source at {sky_coord}"
+            logger.warning(msg)
+
+        return morph, bbox
+
+
+class StarletSource(FactorizedComponent):
+    """Source with a starlet morphology model
+
+    Sources are initialized as `~scarlet.ExtendedSource`, and the morphology is then
+    transformed into starlet coefficients.
+    """
+
+    def __init__(
+        self,
+        model_frame,
+        sky_coord,
+        observations,
+        spectrum=None,
+        thresh=1.0,
+        prerender=False,
+        full=False,
+        starlet_thresh=5e-3,
+    ):
+        """Extended source intialized to match a set of observations
+
+        Parameters
+        ----------
+        model_frame: `~scarlet.Frame`
+            The frame of the model
+        sky_coord: tuple
+            Center of the source
+        observations: instance or list of `~scarlet.observation.Observation`
+            Observation(s) to initialize this source.
+        thresh: `float`
+            Multiple of the backround RMS used as a
+            flux cutoff for morphology initialization.
+        spectrum: `numpy.ndarray` or `scarlet.Parameter`
+            Initial spectrum, otherwise given by `ExtendedSource` initialization
+        full: `bool`
+            If set to False, use the full image as a box for the starlet model.
+            Otherwise, use the same bbox as for `ExtendedSource` initialization
+        starlet_thresh: `float`
+            Multiple of the backround RMS used as a
+            flux cutoff for starlet threshold (usually between 5 and 3).
+        """
+        source = ExtendedSource(
+            model_frame, sky_coord, observations, thresh=thresh, prerender=prerender
+        )
+        source = StarletSource.from_source(source)
+
+        if spectrum is not None:
+            if isinstance(spectrum, Parameter):
+                assert spectrum.name == "spectrum"
+            else:
+                constraint = PositivityConstraint(
+                    zero=1e-20
+                )  # slightly positive values
+                step = partial(relative_step, factor=1e-6)
+                spectrum = Parameter(
+                    spectrum, name="spectrum", step=step, constraint=constraint,
+                )
+            source.children[0] = spectrum
+
+        # still need to init *this* source
+        super().__init__(source.frame, *source.children)
+
+    @classmethod
+    def from_source(cls, source, starlet_thresh=5e-3):
```

Comment:
Sweet, I like this method! A nice way to have the option to initialise a starlet source using an exiisting source of any nature.